File: packages/typescript/src/basic-capabilities/providers/capabilities.ts
import { requireProviderSpec } from "../../generated/spec-helpers.ts";
import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
} from "../../types/index.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("CAPABILITIES");

/**
 * Provider that collects capability descriptions from all registered services
 */
/**
 * Provides capabilities information for the agent.
 *
 * @param {IAgentRuntime} runtime - The agent runtime instance.
 * @param {Memory} _message - The memory message object.
 * @returns {Promise<ProviderResult>} The provider result object containing capabilities information.
 */
export const capabilitiesProvider: Provider = {
  name: spec.name,
  description: spec.description,
  get: async (
    runtime: IAgentRuntime,
    _message: Memory,
  ): Promise<ProviderResult> => {
    // Get all registered services
    const services = runtime.getAllServices();

    if (!services || services.size === 0) {
      return {
        text: "No services are currently registered.",
      };
    }

    // Extract capability descriptions from all services
    const capabilities: string[] = [];

    for (const [serviceType, serviceArray] of services) {
      // Handle the fact that services are stored as arrays
      if (serviceArray && serviceArray.length > 0) {
        // Use the first service in the array for the capability description
        const service = serviceArray[0];
        if (service.capabilityDescription) {
          const agentName = runtime.character.name ?? "Agent";
          capabilities.push(
            `${serviceType} - ${service.capabilityDescription.replace("{{agentName}}", agentName)}`,
          );
        }
      }
    }

    if (capabilities.length === 0) {
      return {
        text: "No capability descriptions found in the registered services.",
      };
    }

    // Format the capabilities into a readable list
    const formattedCapabilities = capabilities.join("\n");

    return {
      data: {
        capabilities,
      },
      text: `# ${runtime.character.name}'s Capabilities\n\n${formattedCapabilities}`,
    };
  },
};

export default capabilitiesProvider;

File: packages/python/elizaos/bootstrap/providers/capabilities.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("CAPABILITIES")


async def get_capabilities(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    model_types = ["TEXT_LARGE", "TEXT_SMALL", "TEXT_EMBEDDING", "IMAGE", "AUDIO"]
    available_models = [mt for mt in model_types if runtime.has_model(mt)]
    service_names = [s.name for s in runtime.services if hasattr(s, "name")]

    features: list[str] = []
    if runtime.get_setting("ENABLE_VOICE"):
        features.append("voice")
    if runtime.get_setting("ENABLE_VISION"):
        features.append("vision")
    if runtime.get_setting("ENABLE_MEMORY"):
        features.append("long_term_memory")

    capabilities: dict[str, list[str] | bool] = {
        "models": available_models,
        "services": service_names,
        "features": features,
    }

    text_parts: list[str] = ["# Agent Capabilities"]

    if available_models:
        text_parts.append(f"Models: {', '.join(available_models)}")

    if service_names:
        text_parts.append(f"Services: {', '.join(service_names)}")

    if features:
        text_parts.append(f"Features: {', '.join(features)}")

    return ProviderResult(
        text="\n".join(text_parts),
        values={
            "modelCount": len(available_models),
            "serviceCount": len(service_names),
            "hasVoice": "voice" in features,
            "hasVision": "vision" in features,
        },
        data=capabilities,
    )


capabilities_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_capabilities,
    dynamic=_spec.get("dynamic", False),
)
